### PR TITLE
Merges DML and other changes made for QLDB

### PR DESCRIPTION
**Pull request by dlurton**
This includes:

- Support for DML statements INSERT, UPDATE, DELETE, etc.
- Support for the BY FROM-source alias
- A preliminary static type checker
- A number of minor bug fixes and refactorings.

Does *not* include DDL statements (CREATE, DROP, and UNDROP).

Contributors to these changes include:  Almann Goo, Rob Calhoun,
David Lurton, Michael Ross and James Bowman.

Commit subject-lines below:

- Add AstNodeTest tests for DML nodes.
- Fix issues remaining after merge
- Merge branch 'quantum-dev' of IonSQLSandbox into quantum-dev
- Merge branch 'mainline' into quantum-dev
- improve ambiguous binding error message (for real this time)
- Revert "Improve ambiguous binding error message"
- Improve ambiguous binding error message
- introduce SelectStarRewriter
- fix LIMIT with value over 2^31 returning no results
- StaticTypeRewriter now requires FromSourceAliasRewriter
- fix variable lookup within FROM sources
- optimize IN operator
- IDs in DML Simple Paths
- Allow UPDATE form to be used for any DML Operation
- add unit tests for CROSSS JOIN with condition and other join types without conditions
- Restricts various JOINs as per PartiQL spec.
- Add support for BY to legacy DML
- Add enhanced validations to StaticTypeRewriter
- Lowercase the id of the count VariableReference
- Capture Token for parse types that recurr on tail
- Support FROM source BY variables in ExprValue and EvaluatingCompliler
- From source BY variable support in parser and (de)serializer.
- Introduces FromSourceLet and LetVariables to the AST
- Add support for V0 to `AstSerializer` and make `V0AstSerializer` a shim
- Add support for DML to StaticTypeRewriter
- Delete zero byte BindingCase.* files which somehow were added to the project root
- Merge in rename of com.amazon.ionsql to org.partiql from mainline
- Add generic type arguments to Bindings references that aren't in the lang project
- Bring quantum-dev in line with 'mainline'
- Reorganize directory structure to match mainline.
- Merge in minor changes to package structure.
- Eliminates MapBindings<T> and fixes StaticTypeRewriter
- Merges in changes to StaticTypeRewriter.

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: docs/sapp-beta.json [hunk 1]
```diff
@@ -0,0 +1 @@
+{"projectName":"IonSQLBeta"}
```
[the remark this answers is not in the document]
Comment (dlurton) [reply docs/sapp-beta.json]: done

File: docs/sapp-release.json [hunk 2]
```diff
@@ -0,0 +1 @@
+{"projectName":"IonSQL"}
```
[the remark this answers is not in the document]
Comment (dlurton) [reply docs/sapp-release.json]: done

File: docs/macros.md [hunk 3]
```diff
@@ -0,0 +1,4 @@
+
```
[the remark this answers is not in the document]
Comment (dlurton) [reply docs/macros.md]: done

File: lang/src/org/partiql/lang/ast/Util.kt [hunk 4]
```diff
@@ -41,7 +41,7 @@ fun createCountStar(ion: IonSystem, metas: MetaContainer): CallAgg {
     }
     return CallAgg(
         funcExpr = VariableReference(
-            id = "COUNT",
```
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/src/org/partiql/lang/ast/Util.kt]: aggregate function names are all stored in lower case now--a cheat on the case-insensitivity thing.

File: lang/src/org/partiql/lang/ast/passes/StaticTypeRewriter.kt [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+            val existing = localsOnlyEnv[BindingName(name, BindingCase.INSENSITIVE)]
+            if (existing != null) {
+                errAmbiguousName(name, metas)
+            }
+            localsMap[name] = type
+            // this requires a new instance because of how [Bindings.ofMap] works
+            localsOnlyEnv = wrapBindings(Bindings.ofMap(localsMap), currentScopeDepth)
+        }
+
+        // TODO https://sim.amazon.com/issues/IONSQL-477
```
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/src/org/partiql/lang/ast/passes/StaticTypeRewriter.kt]: done

File: lang/src/org/partiql/lang/ast/passes/StaticTypeRewriter.kt [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+        override fun rewriteCallAgg(node: CallAgg): ExprNode {
+            return CallAgg(
+                // do not rewrite the funcExpr--as this is a symbolic name in another namespace (AST is over generalized here)
+                node.funcExpr,
+                node.setQuantifier,
+                rewriteExprNode(node.arg),
+                rewriteMetas(node))
+        }
+
+        // TODO https://sim.amazon.com/issues/IONSQL-477
```
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/src/org/partiql/lang/ast/passes/StaticTypeRewriter.kt]: done

File: lang/src/org/partiql/lang/ast/passes/StaticTypeRewriter.kt [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+    /**
+     * With this constraint, all variables from nested queries must resolve to
+     * lexically scoped variables.
+     *
+     * Constraining a user's access to global binds is useful for simplified
+     * query planning in a DB environment where global binds may be costly to
+     * fetch or lead to large intermediate results.
+     */
+    PREVENT_GLOBALS_IN_NESTED_QUERIES;
+}
```
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/src/org/partiql/lang/ast/passes/StaticTypeRewriter.kt]: Yes, `PREVENT_GLOBALS_IN_NESTED_QUERIES` is needed for QLDB, but `PREVENT_GLOBALS_EXCEPT_IN_FROM` is for the general case.

File: lang/src/org/partiql/lang/eval/Bindings.kt [hunk 8]
```diff
@@ -73,70 +81,85 @@ interface Bindings {
...
  *
  * [loweredCaseMap] is based on [originalCaseMap] and is lazily calculated the first time
  * a case-insensitive lookup is performed.
  *
  * If an ambiguous binding match is found during a case-insensitive lookup, [errAmbiguousBinding]
  * is invoked to report the error to the customer.
  */
-public class BindingMap<T>(val originalCaseMap: Map<String, T>) {
-    val loweredCaseMap: Map<String, List<Map.Entry<String, T>>> by lazy {
+class MapBindings<T>(val originalCaseMap: Map<String, T>) : Bindings<T> {
```
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/src/org/partiql/lang/eval/Bindings.kt]: Yes, technically this is a backward-incompatible change and this isn't the only one.  I have not done an A-B test, but we should when we do the next release.  The reason for this change in particular IIRC is that it doesn't handle mutation of `originalCaseMap` very well.

File: lang/src/org/partiql/lang/syntax/LexerConstants.kt [hunk 9]
```diff
@@ -276,8 +278,15 @@ internal val DATE_PART_KEYWORDS: Set<String> = DatePart.values()
     "bag"
 )
 
+/**
+ * QLDB dialect keywords.
+ *
+ * TODO: move from core LexerConstants once there is an extension mechanism
+ */
+@JvmField internal val QLDB_KEYWORDS = setOf("undrop")
+
```
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/src/org/partiql/lang/syntax/LexerConstants.kt]: Removed.

File: lang/test/org/partiql/lang/errors/ParserErrorsTest.kt [hunk 10]
```diff
@@ -834,6 +836,8 @@ class ParserErrorsTest : TestBase() {
                                                 Property.TOKEN_VALUE to ion.newSymbol("as")))
     }
 
+    // FIXME This is still an error--but an error in a different way
+    @Ignore
```
Comment (therapon) [inline lang/test/org/partiql/lang/errors/ParserErrorsTest.kt]: Without blocking this PR, for FIXME's I think we should add issues and reference them in the code.

File: lang/src/org/partiql/lang/ast/passes/StaticTypeRewriter.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/src/org/partiql/lang/ast/passes/StaticTypeRewriter.kt]: @raganhan Arguably, no, but keeping it out of this merge would be more of a pain than allowing it in IMHO.  It's also quite possible that other PartiQL implementations will want the same thing so there's an argument to be keeping it in.

[the remark this answers is not in the document]
Comment (raganhan) [reply lang/src/org/partiql/lang/ast/passes/StaticTypeRewriter.kt]: I agreed that we shouldn't couple this with the merge PR, I'm ok with an issue to investigate this further down the line.

[the remark this answers is not in the document]
Comment (dlurton) [reply lang/src/org/partiql/lang/ast/passes/StaticTypeRewriter.kt]: #158

File: lang/test/org/partiql/lang/errors/ParserErrorsTest.kt (hunk 10 above)
Comment (dlurton) [reply lang/test/org/partiql/lang/errors/ParserErrorsTest.kt]: #159

**Comment by dlurton** [comment]
The only thing we're waiting for to merge this PR is to get the JDK8 build fixed in travis.  It's currently failing due to what I'm 95% sure are problems unrelated to the changes in this PR.

**Comment by therapon** [comment]
> travis JDK8 is still timing out. I'm ok merging anyway
> 
> We should look into removing jdk8 and using jdk11 targeting 8

Ditto.

**Comment by dlurton** [comment]
Ok, I think I finally figured out the source of the gradle `openjdk8` slowness.

It's because [Gradle limits the heap size of JVM forks used for tests to 512mb in total](https://docs.gradle.org/5.0/userguide/upgrading_version_4.html#rel5.0:default_memory_settings).  I think when we added `jacoco` we dramatically increased memory usage of the unit test forks.

It looks like adding:

```
subprojects {

    tasks.withType(Test) {
        maxHeapSize = "4g"
    }
}
```

to the root `build.gradle` may have finally fixed it.  (Phew!)  It's been an inconsistent problem, but I think
this time it's finally will be fixed.

**Pull request merged**